runtime/memprof.c

```diff
@@ -42,1290 +563,1749 @@ struct memprof_thread_s {
-  double saved_lambda = lambda;
+  value res = Val_unit;
+  CAMLlocal1(config);
+  config = entries->config;
 
-  /* If this condition is false, then [caml_memprof_young_trigger] should be
-     equal to [Caml_state->young_alloc_start]. But this function is only
-     called with [Caml_state->young_alloc_start <= Caml_state->young_ptr <
-     caml_memprof_young_trigger], which is contradictory. */
-  CAMLassert(!local->suspended && lambda > 0);
+  /* When a domain is not sampling, the memprof trigger is not
+   * set, so we should not come into this function. */
+  CAMLassert(sampling(domain));
 
   if (!from_caml) {
-    unsigned n_samples = 1 +
-      rand_binom(caml_memprof_young_trigger - 1 - Caml_state->young_ptr);
-    CAMLassert(encoded_alloc_lens == NULL);    /* No Comballoc in C! */
-    caml_memprof_renew_minor_sample();
-    maybe_track_block(Val_hp(Caml_state->young_ptr), n_samples,
-                      wosize, SRC_NORMAL);
-    return;
+    /* Not coming from Caml, so this isn't a comballoc. We know we're
+     * sampling at least once, but maybe more than once. */
+    size_t samples = 1 +
+      rand_binom(domain,
+                 Caml_state->memprof_young_trigger - 1 - Caml_state->young_ptr);
+    CAMLassert(encoded_lens == NULL);
+    maybe_track_block(domain, Val_hp(Caml_state->young_ptr),
+                      samples, wosize, CAML_MEMPROF_SRC_NORMAL);
+    caml_memprof_set_trigger(Caml_state);
+    caml_reset_young_limit(Caml_state);
+    CAMLreturn0;
   }
 
-  /* We need to call the callbacks for this sampled block. Since each
-     callback can potentially allocate, the sampled block will *not*
-     be the one pointed to by [caml_memprof_young_trigger]. Instead,
-     we remember that we need to sample the next allocated word,
-     call the callback and use as a sample the block which will be
-     allocated right after the callback. */
-
-  CAMLassert(Caml_state->young_ptr < caml_memprof_young_trigger &&
-             caml_memprof_young_trigger <= Caml_state->young_ptr + whsize);
-  trigger_ofs = caml_memprof_young_trigger - Caml_state->young_ptr;
-  alloc_ofs = whsize;
-
-  /* Restore the minor heap in a valid state for calling the callbacks.
-     We should not call the GC before these two instructions. */
+  /* The memprof trigger lies in (young_ptr, young_ptr + whsize] */
+  CAMLassert(Caml_state->young_ptr < Caml_state->memprof_young_trigger &&
+             Caml_state->memprof_young_trigger <=
+               Caml_state->young_ptr + whsize);
+
+  /* Trigger offset from the base of the combined allocation. We
+   * reduce this for each sample in this comballoc. Signed so it can
+   * go negative. */
+  intnat trigger_ofs =
+    Caml_state->memprof_young_trigger - Caml_state->young_ptr;
+  /* Sub-allocation offset from the base of the combined
+   * allocation. Signed so we can compare correctly against
+   * trigger_ofs. */
+  intnat alloc_ofs = whsize;
+
+  size_t new_entries = 0;
```

Comment:
```suggestion
#ifdef DEBUG
  size_t new_entries = 0;
#endif
```

Reply:
My alternative is to add a null usage `(void)new_entries;`, to tell the compiler not to warn.